src/Workspaces/Core/Portable/Remote/RemoteHostClient.cs

```diff
@@ -104,35 +62,57 @@ private void OnConnectionChanged(bool connected)
             ConnectionChanged?.Invoke(this, connected);
         }
 
-        private async Task<PinnedRemotableDataScope> GetPinnedScopeAsync(Solution solution, CancellationToken cancellationToken)
+        /// <summary>
+        /// NoOpClient is used if a user killed our remote host process. Basically this client never
+        /// create a session
+        /// </summary>
+        public class NoOpClient : RemoteHostClient
         {
-            if (solution == null)
+            public NoOpClient(Workspace workspace) :
+                base(workspace)
+            {
+            }
+
+            public override Task<Session> TryCreateSessionAsync(string serviceName, object callbackTarget, CancellationToken cancellationToken)
             {
-                return null;
+                return SpecializedTasks.Default<Session>();
             }
 
-            Contract.ThrowIfFalse(solution.Workspace == _workspace);
+            protected override void OnConnected()
+            {
+                // do nothing
+            }
 
-            var service = _workspace.Services.GetService<ISolutionSynchronizationService>();
-            return await service.CreatePinnedRemotableDataScopeAsync(solution, cancellationToken).ConfigureAwait(false);
+            protected override void OnDisconnected()
+            {
+                // do nothing
+            }
         }
 
-        // TODO: make this to not exposed to caller. abstract all of these under Request and Response mechanism
         public abstract class Session : IDisposable
         {
-            protected readonly PinnedRemotableDataScope PinnedScopeOpt;
             protected readonly CancellationToken CancellationToken;
+            private PinnedRemotableDataScope _scope;
 
             private bool _disposed;
 
-            protected Session(PinnedRemotableDataScope scope, CancellationToken cancellationToken)
+            protected Session(CancellationToken cancellationToken)
             {
                 _disposed = false;
+                _scope = null;
 
-                PinnedScopeOpt = scope;
                 CancellationToken = cancellationToken;
             }
 
+            public PinnedRemotableDataScope PinnedRemotableDataScope => _scope;
```

Comment:
> all these I hope to get rid of.

It would be helpful to get a sketch of what the API is (i.e. interfaces only) that you want to morph this all into.  I'm not really clear on where you're trying to get to and it would be helpful to the discussion if more information was provided here.

If this is a stepping stone to a larger amount of refactoring, then i'd be more amenable to it.  But i really want to know what you envision the final shape being so that i can feel comfortable with these intermediary steps.